### ISSUE TITLE
## New Feature/Enhancement- Multi items price

### ISSUE THREAD

**Issue by beinafuushop** [jekkos/opensourcepos#1017]
### New Feature / Enhacement

Hey Team,

Am thinking if you can add price tiers for items?such as multiple price tiers can be configured for all items and while selling,you select which price tier to use ,Am expecting the community to develop the feature but i will donate $50.

**Comment by beinafuushop** [comment]
Team,

Any feedback in this?

**Comment by chandrawg** [comment]
subscribe

**Comment by alvin101** [comment]
Price sheets that you can set and then pin to the products.  It should also match a popular erp system.  

I will throw in a little money...

**Comment by beinafuushop** [comment]
@daN4cat, can you please let us know what is needed for community to assist develop this feature?Cause this is very important for business with product thay has multiple price tires.

**Comment by jekkos** [comment]
This sounds like a rather complicated addition. Also requirements are pretty unclear. On what do you want to base price setting? If we need to come up with some kind of generic rule based system then I think it will be quite a lot of work.

**Comment by jekkos** [comment]
One will probably need to decouple prices from items and store them in a separate table. Then after evaluating some kind of rule the correct price can be determined. I guess one would also like some kind of configuration screen to create and maintain pricing rules.

**Comment by jlctmaster** [comment]
Hi, for this that @beinafuushop requires you should create a price list structure, where you can manage by its different versions according to a date of publication, and apply all the modifications to the sales module to take the price of the corresponding list according The system current date.

It is a strong job, but it would give the system a lot of functionality.

Greetings.

**Comment by beinafuushop** [comment]
From my side, what am proposing is to have more than one selling price for a single items, for example retail price and whole price for each item, in the sale page, there can be a price category option just like the way we have mode for sale or return, then from price category two option will be available, for retail price and whole price, depending on the price category set,all the items that will be inserted in cart will pick the corresponding selling price.

**Comment by jlctmaster** [comment]
Hello @beinafuushop 

For what you request, I suggest to manage a module of promotions, where you can manage to product to single price, but in this module you can condition X amount of promotions for that item, in your case if it is to 'wholesale 'Side Of' X 'quantity of products you apply' N 'discount percentage.

Greetings.

**Comment by beinafuushop** [comment]
Hello @jlctmaster,

some time it is a challenge to use discount option to manage price, cause each item will have its own discount percent, Now if you have multiple items then things will be more difficult.

That is why i was thinking of having price tiers,where for each item you can define more than one selling price and while selling, you decide which tier to use.

**Comment by jlctmaster** [comment]
Could you give me an example of any billing where that method is used? To try to better understand your requirement and see what solution could be applied there.

Greetings.

**Comment by beinafuushop** [comment]
Greetings,

I don't have a specific billing system to refer to, though i can draw item and sale page to reflect what am thinking, and share the screen shots here, am not sure if it will be of more valuable for required clarification.

**Comment by jlctmaster** [comment]
Of course, anything serves to clarify the doubts with respect to your request.

It can be a document in excel where you show the product with its different prices, to verify what you require.

Greetings.

**Comment by alvin101** [comment]
You have a customer that buys wholesale. To this customer product "A" is discounted 10% if he buys a gross of "A" he gets a 20% discount, Product "B" is discounted 30%

This is all done automatically from price sheets that are set to products and customers. 

And yes it would look a lot like a spreadsheet.

**Comment by jlctmaster** [comment]
Ok, with the example you are mounting, applying a parameterizable promotions module to the system can work. for example:

1- A promo teacher
2- a window of pre-conditions: where you can parameterize the client, promo codes, promotion dates by rank, etc.
3- a distribution window: where you can parameterize the different conditions that can be applied for example: product A buys 15 units.
4- a window of prizes: where you can parameterize the prizes to be applied if it meets the conditions previously defined, for example: percentages, amount, etc.

Greetings.

**Comment by beinafuushop** [comment]
Hello Team,

As discussed, attached is the excel sheet with the clarification in the requirement, Please let me know if it is still not clear.
Cheers.
[OSPOS_New_Feature_Requirement_Price_Tiers.xlsx](https://github.com/jekkos/opensourcepos/files/648450/OSPOS_New_Feature_Requirement_Price_Tiers.xlsx)

**Comment by beinafuushop** [comment]
Hello Team,

Any inputs?

**Comment by jlctmaster** [comment]
Hello @beinafuushop 

I have reviewed the excel file that you shared and for that you would have to take a control of the product to be able to distinguish each customer and be able to parameterize it to define what price it is going to 
be invoiced.

**Comment by beinafuushop** [comment]
Hello jlctmaster,

Yes that is the case, but for this service seller must have separate business terms used to define selling price for each customer,
Now this is understood, is it feasible for.implementation?can community consider this requirement and share plan for implementation and pre requisites?

**Comment by alvin101** [comment]
My two cents,

I still think this kind of feature belongs in the back office, POS should grab data from there when needed.

**Comment by jlctmaster** [comment]
Hello,

I think that if it should be a functionality within the POS as in many trades handle this type of transaction where for the same product have more than one mode of sale, retail, wholesale and by promotion. I suggest that for the case of wholesale and retail, you can manage discount schemes configurable by product and applicable in the master of customers, and for promotions a complete module to this process.

**Comment by beinafuushop** [comment]
Hello jlctmaster,

I don"t understand what you are suggesting, do we have a feature for.management of promositions in the latest version (V3.0.1)?,If yes then how to use it?

Regarding using discount for different price management, this option is challenge assume you have a 1000 product and each product has its own retail and whole price, even if you calculate the percent still you can not have same discount percent to apply for all items.

**Comment by jlctmaster** [comment]
No, the system does not have a promotions module, it is only a suggestion to apply it, with respect to the discount schemes, it is manageable, suppose the business is a wholesale business, where it sells wholesale and retail products , Then that business has different types of customers, the final consumers who buy the retail, and the small businesses that buy by boxes of products and therefore they sell him wholesale, that you can implement it with a discount scheme that is Applicable to a group of customers or using the discount field of the customer master that already owns the system. Only at the moment of modifying that % discount you must apply it client by customer.

**Comment by beinafuushop** [comment]
The problem of discount scheme comes when there are many items of different prices, it means you can not apply same dis ount percent, you will need to have separate discount percent for each item, which is great challenge.to keep/memorize percent discount for all items.

**Comment by beinafuushop** [comment]
But if you have a room to configure different price tiers (as shown in attached excel above) it will be great, cause you will set appropriate tier based on customer you are selling to.

Please team consider this as option and share your inputs.

**Comment by beinafuushop** [comment]
Hello Team 

Please give the input and way foward.

**Comment by jlctmaster** [comment]
Hello, to do it as you suggest in the excel sheet would only lack one variable.

At what time or under which parameterization the system is told to apply the retail or wholesale price.

Greetings.

**Comment by beinafuushop** [comment]
Hello,

I think in the current version, there is parameter which insert retail price in the cart by the time item selected, Same parameter can be used I think, but this time it should not pick retail price by  default, there should be a logic to instruct parameter which selling price to take depend on what is current sated in the price category.

**Comment by jlctmaster** [comment]
Exactly, there must be another control form, either by telling the system by type of customer, whether they are wholesalers or not, for example.

**Comment by beinafuushop** [comment]
Hello,

Now that you understand the requirement, can you please let me know if this requirement is feasible?and what is the next step?

**Comment by jlctmaster** [comment]
Hi @beinafuushop , for me the requirement is feasible, in fact I can make modifications to the system to support it, however, it requires the approval of @jekkos or @daN4cat to work on this and contribute in the main branch, otherwise it would be an exclusive modification That you would do to a fork of that project.

**Comment by beinafuushop** [comment]
Hello jlctmaster,

@jekkos and/or @daN4cat, can you please provide your inputs?

**Comment by SteveIreland** [comment]
This looks like it would be a nice feature to have.  If I understood the key points of the thread -  I understand that the price for an item would be based on a price category assigned at a customer level (with possible override at sale time) and it would be based on the current system date, which suggests that I can set up future price changes that would be effective as of a given date.

**Comment by beinafuushop** [comment]
Hello @steveIreland,

Yes some thing like that, but what is important is to have a price tiers for all items and selected based on customer.

**Comment by jekkos** [comment]
@jlctmaster Sure you are welcome to add this part to the system. You could of course start with a new branch in your own repository and send us a pull request when finished. This makes it easy for us to review what you did and discuss about changes that might  be needed.

I think thi is certainly not an easy modification and it requires some thoughtwork to get it right. Therefore maybe we should discuss a bit beforehand to see where we can help with the implementation.

**Comment by jlctmaster** [comment]
Ok, we can discuss the modifications that should be applied to support this improvement, can we do it here? Or what is the procedure for the case?

**Comment by jekkos** [comment]
Yes sure, take a look in the design/ directory. There's some ERD in there and if you have database changes, might be cool to add a model to the diagram and post it here so we can discuss a bit.

**Comment by jekkos** [comment]
I'll try to see if I can come up with some ideas for this one, however I don't have the urge to implement it myself. 

I guess one could go solve this by adding an intermediate table between items and item_prices, called item_price_rules or something. This table will then contain some columns holding the parameters of the rule you want to evaluate. A business logic inside PHP can then query those three tables and resolve a correct price depending on the parameters given.

The UI part is another story then and it's layout will depend on the needed features. A separate module might be built that allows one to do CRUD on the pricing rules. Prices linked to sales and receivings have to remain immutable. On the other hand prices linked to items will need to be resolved depending on parameters at hand

**Comment by jlctmaster** [comment]
OK sure! Let me evaluate the model entity relationship and propose some ideas there to work on that.

**Comment by jekkos** [comment]
Ok sounds great.

**Comment by bkrajendra** [comment]
This will be really great to have enhancement. I do requested similar earlier #642 . I just came across this discussion and find it useful.

**Comment by bkrajendra** [comment]
I have a thought in my mind, I don't know if its really a practical but just wanted to throw.

I have a situation of two price tags for one item as I'm purchasing Item from two different suppliers.
Now if I get a product with any change - Tax, Purchase Cost or selling cost may be from same supplier or two different supplier, technically its totally 2 different products for me even if it has same barcode.

So my silly thought is : Is it possible if we can make barcode as not Unique.
So in this case 

-  if all three parameters like p.cost, s.cost and tax are same, existing Item will be updated with Inventory.
- but if any thing out of these 3 parameter changes (while receiving), will create new Item with same barcode.

Now on sales screen if we get single entry for barcode search, item will be directly added to cart.
But if we get more than one search result for barcode, it will show us a popup screen with search result showing choice of  item. We can even mark these item with some color, so that  stock can be distinguished. Same color tag can be added to shelf of these Item.

is this really a proper way from accounting point of view, or i'm just talking nonsense.?

**Comment by jlctmaster** [comment]
In my experience working with administrative software, I suggest that the most appropriate way to handle these cases in the system is through the use of a price list that maintains a valid time, and that the product is coded in batches where you can have the same Product but with lots where the price is different, in addition to this you can handle a FiFo rule in the system configuration.

**Comment by jekkos** [comment]
Best way to solve this is encode some extra attribute in the barcode and then decode and select based on barcode+attribute it when scanning.. so barcode remains unique. At least this would be the easiest way to implement it.

**Comment by izeblood** [comment]
@beinafuushop you can use stock location... set stock as retail and add stock2 and rename it to wholesale..Woahhlahhhh... it solve your problem...

**Comment by izeblood** [comment]
tagging from #1585 #1651 Multiple prices for a product is already solved by @SteveIreland when he develops multi pack feature... in which you can easily identify that the product is for retail or for wholesale why? because in the end of the product name it indicates | Each, | Pcs. , --> for retail and 
| Cartoon, |Pack, | Case --> for wholesale ...

**Comment by bkrajendra** [comment]
@izeblood, I have not checked this feature. I'll will take a look at it.
But understand that, multi price simply means even if its barcode is same it should have multiple options, it can be price or attribute (size, weight).
So when we scan a barcode for a product, instead of directly adding it to cart, it should show a popup with options to select. 
Internally it can be 4 different products (with change in price, size, weight etc) but with single barcode.

**Comment by izeblood** [comment]
> @izeblood, I have not checked this feature. I'll will take a look at it.
> But understand that, multi price simply means even if its barcode is same it should have multiple options, it can be price or attribute (size, weight).
> So when we scan a barcode for a product, instead of directly adding it to cart, it should show a popup with options to select.
> Internally it can be 4 different products (with change in price, size, weight etc) but with single barcode.

I think that method will be optional and be added to the configuration settings. That would be great.

**Comment by izeblood** [comment]
@bkrajendra any updates to this enhancement....?

**Comment by izeblood** [comment]
@bkrajendra  HOW ABOUT THIS:
![1](https://user-images.githubusercontent.com/32444665/74394302-4ece2a00-4e47-11ea-8f51-ce2a8b2db09a.jpg)
![2](https://user-images.githubusercontent.com/32444665/74394546-c69c5480-4e47-11ea-987b-27f74b0f38e1.jpg)

**Comment by izeblood** [comment]
![3](https://user-images.githubusercontent.com/32444665/74394712-488c7d80-4e48-11ea-872d-09cf0f15c580.jpg)

**Comment by izeblood** [comment]
THAT'S ONLY A PRESENTATION in order to visualize additional feature for ospos

**Comment by andriux1990** [comment]
@izeblood looks very interesting

**Comment by bkrajendra** [comment]
@izeblood looks good.
But this will only work if we are typing product item name.
What if we are scanning an item with barcode reader. it immediately puts entire barcode into text box and sends ENTER. and products gets added.
So we need to show a popup incase it has more than one prices, else just add to cart.

We need to handle two things.
2) Same Barcode but two different items (Item might be same but with different attributes.)
More technically what if we have barcode table and Items table as separate tables.
Also there should be One-to-Many relationships between barcode and Items.
```
[PSEUDOCODE]
barcode->hasMany(['items'=>itemModel])
```
this way we can maintain inventories and also fetch multiple items by same barcode.
I understand it can be done by existing tricks (stocks or Kits), but we need to think about 
real solution, the way it's handled in commercial POS.

3) GUI - how to show search screen and add screen UI we need to thing about.

ORM and `Eloquent `models are very well handled in `Laravel `where these things can be done easily.

**Comment by izeblood** [comment]
yes.. your right @bkrajendra

**Comment by andriux1990** [comment]
> @bkrajendra HOW ABOUT THIS:
> ![1](https://user-images.githubusercontent.com/32444665/74394302-4ece2a00-4e47-11ea-8f51-ce2a8b2db09a.jpg)
> ![2](https://user-images.githubusercontent.com/32444665/74394546-c69c5480-4e47-11ea-987b-27f74b0f38e1.jpg)

@izeblood do you have all these changes?

**Comment by bkrajendra** [comment]
@andriux1990 this really looks close. WIll have to test it.
Is it on PR?
I hope these multi cost items are reflected in inventory properly.

**Comment by andriux1990** [comment]
@bkrajendra @izeblood This seems very good to me, it can be used in several ways, if you have the code, it would be great to share so you can try it

**Comment by DamianoSilverhand** [comment]
@bkrajendra @izeblood @jlctmaster  I am also trying to implement Multi-pricing dependent on attributes, for instance I have a type of T-shirt whose price depends on the size selected and will appreciate if you have any ideas on how to work around it.

**Comment by ozzylamberth** [comment]
I think that assigned a column in the ospos_items table of the database between cost_price and unit_price we put wholesale_price and through a dropdown btn put the 2 options that when selecting option 1 loads unit_price of any product and if you select option 2 load all the wholesale of each product

Creo que asignado una columna en la tabla ospos_items de la base de datos entre cost_price y unit_price colocamos wholesale_price y por medio de un btn dropdown colocar las 2 opciones que al seleccionar opcion 1 cargue unit_price de cualquier producto y si selecciona la opcion 2 cargue todos los wholesale de cada producto

**Comment by loejo** [comment]
> > @bkrajendra HOW ABOUT THIS:
> > ![1](https://user-images.githubusercontent.com/32444665/74394302-4ece2a00-4e47-11ea-8f51-ce2a8b2db09a.jpg)
> > ![2](https://user-images.githubusercontent.com/32444665/74394546-c69c5480-4e47-11ea-987b-27f74b0f38e1.jpg)
> 
> @izeblood do you have all these changes?

please i need that how to do it ???

**Comment by odiea** [comment]
https://github.com/opensourcepos/opensourcepos/issues/2989

**Comment by Lankage02** [comment]
> > @bkrajendra HOW ABOUT THIS:
> > ![1](https://user-images.githubusercontent.com/32444665/74394302-4ece2a00-4e47-11ea-8f51-ce2a8b2db09a.jpg)
> > ![2](https://user-images.githubusercontent.com/32444665/74394546-c69c5480-4e47-11ea-987b-27f74b0f38e1.jpg)
> 
> @izeblood do you have all these changes?

can me get a this souce code...

**Comment by mainamwangi2023** [comment]
can you share the code


> > > @bkrajendra HOW ABOUT THIS:
> > > ![1](https://user-images.githubusercontent.com/32444665/74394302-4ece2a00-4e47-11ea-8f51-ce2a8b2db09a.jpg)
> > > ![2](https://user-images.githubusercontent.com/32444665/74394546-c69c5480-4e47-11ea-987b-27f74b0f38e1.jpg)
> > 
> > 
> > @izeblood do you have all these changes?
> 
> can me get a this souce code...

share the code please

**Comment by Hallsie** [comment]
> From my side, what am proposing is to have more than one selling price for a single items, for example retail price and whole price for each item, in the sale page, there can be a price category option just like the way we have mode for sale or return, then from price category two option will be available, for retail price and whole price, depending on the price category set,all the items that will be inserted in cart will pick the corresponding selling price.

I would think that the more elegant way to do it would be to use the Customer feature. You would set the customer which would define your taxes as well as "wholesale" or "retail" and then obviously you can still have your discount pricing attached as well at that point in time. So if you have a wholesale customer that gets 10% off then it would still do that.

On item creation you would have then: Cost, Wholesale, and Retail pricing. I'm not sure how you would handle bulk pricing, that would be separate.

Then the customer would have to tell you up front if it is a wholesale or retail purchase they are making and either you can set two separate customers or you would have to have a whole separate way to determine if you are doing wholesale vs. retail for one customer that falls under both.
